lib/index.js

```diff
@@ -29,59 +57,57 @@ export default function runContentfulExport (usageParams) {
   })
     .then((response) => {
-      if (opts.downloadAssets) {
-        var successCount = 0
-        var warningCount = 0
-        var errorCount = 0
+      if (!opts.downloadAssets) {
+        return response
+      }
+      let successCount = 0
+      let warningCount = 0
+      let errorCount = 0
 
-        log.info('Downloading ' + response.assets.length + ' assets')
+      log.info('Downloading ' + response.assets.length + ' assets')
 
-        var arrayOfDownloadPromises = []
-        response.assets.forEach((asset) => {
-          if (asset.fields.file) {
-            Object.keys(asset.fields.file).forEach((key) => {
-              arrayOfDownloadPromises.push(
-                downloadAsset(asset.fields.file[key].url, opts.exportDir).then(
-                  (downLoadedFile) => {
-                    log.info('-> ' + downLoadedFile)
-                    successCount++
-                  })
-                  .catch((error) => {
-                    const {url} = asset.fields.file[key]
-                    log.error('-> error downloading ' + url + ' => ' + error)
-                    errorCount++
-                    return Promise.resolve(url)
-                  }
-                )
-              )
-            })
-          } else {
-            log.warn('-> asset has no file(s)', jsonStringifySafe(asset))
-            warningCount++
-          }
+      return Promise.map(response.assets, (asset) => {
+        if (!asset.fields.file) {
+          log.warn('-> asset has no file(s)', jsonStringifySafe(asset))
+          warningCount++
+          return
+        }
+        const locales = Object.keys(asset.fields.file)
+        return Promise.mapSeries(locales, (locale) => {
+          const url = asset.fields.file[locale].url || asset.fields.file[locale].upload
+          return downloadAsset(url, opts.exportDir)
+          .then((downLoadedFile) => {
+            log.info('-> ' + downLoadedFile)
+            successCount++
+          })
+          .catch((error) => {
+            log.error('-> error downloading ' + url + ' => ' + error.message)
+            log.error(JSON.stringify(error, null, 2))
+            errorCount++
+          })
         })
-
-        Promise.all(arrayOfDownloadPromises).then(
-          () => {
-            log.info('All ' + response.assets.length + ' downloads finished')
-            log.info('successes: ' + successCount)
-            log.info('warnings: ' + warningCount)
-            log.info('errors: ' + errorCount)
-          }
-        )
-      }
-      if (exportToFile) {
-        fs.existsSync(opts.exportDir) || fs.mkdirSync(opts.exportDir)
-        const responseFile = `${opts.exportDir}/contentful-export-${clients.source.spaceId}-${Date.now()}.json`
-        log.info('Writing space data to json file at : ' + responseFile)
-        return fs.writeFile(responseFile, jsonStringifySafe(response, null, 4))
-      }
-      return response
+      }, {
+        concurrency: 10
```

Comment:
Hm, @Khaledgarbaya why reduce it? S3 has concurrency limits higher than 10 req/s